File: getting_started/grpc.cpp
#include <format>
#include <fstream>
#include <grpcpp/grpcpp.h>
#include <jwt-cpp/jwt.h>

#include "roguedb/roguedb.grpc.pb.h"
#include "roguedb/queries.pb.h"
#include "roguedb/test.pb.h"

std::string createJwt()
{
    // Values found in service_account.json.
    const std::string SERVICE_ACCOUNT_EMAIL{ "YOUR_SERVICE_ACCOUNT_EMAIL" };
    const std::string PRIVATE_KEY_ID{ "YOUR_PRIVATE_KEY_ID" };
    const std::string PRIVATE_KEY{ "YOUR_PRIVATE_KEY" };
    const auto now{ std::chrono::system_clock::now() };
    return std::string{ jwt::create()
        .set_issuer(SERVICE_ACCOUNT_EMAIL)
        .set_subject(SERVICE_ACCOUNT_EMAIL)
        .set_audience(std::format(
        	"{}.roguedb.dev", 
            SERVICE_ACCOUNT_EMAIL.substr(
            	0, SERVICE_ACCOUNT_EMAIL.find("@"))))
        .set_issued_at(now)
        .set_expires_at(now + std::chrono::hours(1))
        .set_header("kid", PRIVATE_KEY_ID)
        .sign(jwt::algorithm::rs256{ PRIVATE_KEY }) };
}

std::vector<std::filesystem::path> detectFiles(
    const std::vector<std::string>& directories)
{
    std::vector<std::filesystem::path> files{};
    for(const auto& directory : directories)
    {
        for(const auto& entry : std::filesystem::directory_iterator{directory})
        {
            if(!std::filesystem::is_directory(entry))
            {
                if(entry.path().extension() == ".proto")
                {
                    files.emplace_back(std::move(entry.path()));
                }
            }
            else
            {
                std::vector<std::string> temp{ entry.path() };
                for(auto& filename : detectFiles(entry.path()))
                {
                    files.emplace_back(std::move(filename));
                }
            }
        }
    }
    return files;
}

int main(int argc, char** argv)
{
    // See purchase confirmation emails for details and service_account.json.
	const std::string API_KEY{ "YOUR_API_KEY" };
    const std::string URL{ "c-[YOUR_IDENTIFIER_FIRST_28_CHARACTERS].roguedb.dev" };
    const std::string ENCODED_JWT{ createJwt() };
    
    ////////////////////////////////////////////////////////
    ///////  Insert, Update, and Remove API Example  ///////
    ////////////////////////////////////////////////////////
    
    rogue::utilities::Test test{};
    test.set_attribute1(10);
    
    rogue::services::Insert request{}; // Insert API
    // rogue::services::Update update{}; // Update API
    // rogue::services::Remove remove{}; // Remove API
    
    request.set_api_key(API_KEY);

    // Insert, Update, and Remove are identical in use.
    request.add_messages()->PackFrom(test);
    
    // gRPC Connection
    std::unique_ptr<rogue::services::RogueDB::Stub> stub{ rogue::services::RogueDB::NewStub(
    grpc::CreateCustomChannel(
        std::format("{}:443", URL),
        grpc::SslCredentials(grpc::SslCredentialsOptions()))) };
        
    grpc::ClientContext context{};
    context.AddMetadata("Authorization", std::format("Bearer {}", ENCODED_JWT));
    auto stream{ stub->insert(&context) };
    // auto stream{ stub->update(&context) }; // Update API
    // auto stream{ stub->remove(&context) }; // Remove API

    stream->Write(request);
    stream->WritesDone(); // Signal all queries sent. Otherwise, blocks.
    
    // No response is given for Insert, Update, and Remove
    // Any errors get reported in status.
    grpc::Status status{ stream->Finish() };
    
    //////////////////////////////////////
    ////////  Search API Example  ////////
    //////////////////////////////////////

    {
        // Example of a baisc index query.
        rogue::services::Search search{};
        search.set_api_key(API_KEY);
        
        // For Test, attribute1, attribute2, and attribute3 form the index.
        // Search Query:
        // Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
        // AND
        // Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
        rogue::services::Basic& expression{ *search.add_queries()->mutable_basic() };
        expression.set_logical_operator(rogue::services::LogicalOperator::AND);

        test = rogue::utilities::Test{};
        expression.add_comparisons(rogue::services::ComparisonOperator::GREATER_EQUAL);
        test.set_attribute1(1);
        (*expression.add_operands()).PackFrom(test);
            
        expression.add_comparisons(rogue::services::ComparisonOperator::LESSER_EQUAL);
        test.set_attribute1(10);
        (*expression.add_operands()).PackFrom(test);

        grpc::ClientContext context{};
        rogue::services::Response response{};
        std::shared_ptr<grpc::ClientReaderWriter<
            rogue::services::Search, rogue::services::Response>> stream{
                roguedb->search(&context) };
            
        stream->Write(search);
        stream->WritesDone(); // Signal all queries sent. Otherwise, blocks.

        // Multiple queries or large queries should be processed on a separate thread
        // to prevent blocking the database.
        stream->Read(&response);
        status = stream->Finish(); // Signal all queries sent. Blocks otherwise.

        std::vector<rogue::utilities::Test> results{};
        // Queries are zero-indexed. Partial results get sent
        // and mapped to that index.
        for(const auto& result : response.results()[0].messages())
        {
            result.PackTo(results.emplace_back());
        }

        // Each response sends a list of the query ids finished
        // with processing.
        for(const auto& finishedId : response.finished())
        {
            if(finishedId == 0)
            {}
        }
    }
    {
        // Example of a baisc non-indexed query.
        rogue::services::Search search{};
        search.set_api_key(API_KEY);
        
        // Search Query: 
        // Test.attribute1 >= 1 && Test.attribute2 <= 10
        rogue::services::Basic& expression{ *search.add_queries()->mutable_basic() };
        expression.set_logical_operator(rogue::services::LogicalOperator::AND);

        test = rogue::utilities::Test{};
        test.set_attribute1(1);
        expression.add_fields(1); // Corresponds to field id in test.proto
        (*expression.add_operands()).PackFrom(test);
        expression.add_comparisons(rogue::services::ComparisonOperator::GREATER_EQUAL);
            
        test.set_attribute2(10);
        expression.add_fields(2); // Corresponds to field id in test.proto
        (*expression.add_operands()).PackFrom(test);
        expression.add_comparisons(rogue::services::ComparisonOperator::LESSER_EQUAL);
    }

    ///////////////////////////////////
    //// Schema Change API Example ////
    ///////////////////////////////////

    rogue::services::Subscribe subscribe{};
    subscribe.set_api_key(apiKey);

    const std::vector<std::string> directories{ 
        "absolute/path/to/protos/directory1",
        "absolute/path/to/protos/directory2" };

    // All proto files should be sent in a list of
    // their contents. No modifications required.
    for(const auto& file : detectFiles(directories))
    {
        std::ifstream inputFile{ file.native() };
        std::stringstream buffer{};
        buffer << inputFile.rdbuf();
        subscribe.add_queries(buffer.str());
    }

    // Any schemas excluded will have associated data deleted.
    // Schema change failure results in no changes applied.
    roguedb.subscribe(&context, subscribe, &response, metadata=metadata);
}

File: getting_started/grpc.py
import datetime
import json
import jwt
import time
import os
from pathlib import Path

# gRPC and Protocol Buffers Imports
import grpc
from grpc_status import rpc_status

from roguedb.roguedb_pb2_grpc import RogueDBStub
from roguedb.queries_pb2 import (
    Insert, Update, Remove, Basic, Search, 
    LogicalOperator, ComparisonOperator, Subscribe)
from roguedb.test_pb2 import Test

# External Dependency Requirements:
# grpcio, protobuf, PyJWT

def create_jwt(service_account: str, expire_minutes: int = 60) -> str:
    # NOTE: Replace with file path to service_account.json
    # from purchase confirmation email.
    with open(service_account) as input_file:
        key_data = json.load(input_file)

    now = datetime.datetime.now(datetime.timezone.utc)
    iat = int(time.mktime(now.timetuple()))
    exp = int(time.mktime((
        now + datetime.timedelta(minutes=expire_minutes)).timetuple()))

    payload = {
        "iat": iat,
        "exp": exp,
        "iss": key_data['client_email'],
        "aud": f"{key_data['client_email'].split('@')[0]}.roguedb.dev", 
        "sub": key_data['client_email'] }

    headers = {
        "kid": key_data['private_key_id'],
        "alg": "RS256", # Google Cloud service accounts typically use RS256
        "typ": "JWT" }

    return jwt.encode(
        payload, key_data['private_key'], 
        algorithm="RS256", headers=headers)

def detect_files(directories: list[str]) -> list[Path]:
    files = []
    for directory in directories:
        with os.scandir(directory) as scan:
            for item in scan:
                path = Path(item.path)
                if path.is_file() and path.suffix == ".proto":
                    files.append(path)
                elif path.is_dir():
                    for subitem in detect_files(directory=[item.path, ]):
                        files.append(subitem)
    return files

if __name__ == "__main__":
	# See purchase confirmation emails for details and service_account.json.
    url = "c-[YOUR_IDENTIFIER_FIRST_28_CHARACTERS].roguedb.dev"
    api_key = "YOUR_API_KEY"
    encoded_jwt = create_jwt("path/to/service_account.json")

    ################################################
    #### Insert, Update, and Remove API Example ####
    ################################################
    
    request = Insert(api_key=api_key) # Insert API
    # update = Update(api_key=api_key) # Update API
    # remove = Remove(api_key=api_key) # Remove API
    
    # Insert, Update, and Remove are identical in use.
    request.messages.add()

    # See test.proto for the schema definition.
    test = Test(attribute1=10)
    request.messages[0].Pack(test)
    
    roguedb = RogueDBStub(
        channel=grpc.secure_channel(f"{url}:443", 
        grpc.ssl_channel_credentials()))

    def yield_request():
        yield request

    try:
        for _ in roguedb.insert(
        # for _ in roguedb.update( # Update API call
        # for _ in roguedb.remove( # Remove API call
            yield_request(), 
            metadata=[("authorization", f"Bearer {encoded_jwt}")]):
            pass
    except grpc.RpcError as rpc_error:
        # Any errors get reported in status.
        status = rpc_status.from_call(rpc_error)
    
    ###############################
    ##### Search API Examples #####
    ###############################

    # Example of a basic index query. 
    # For Test, attribute1, attribute2, and attribute3 form the index.
    # Search Query: 
    # Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
    # AND
    # Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
    search = Search(api_key=api_key)
    expression = search.queries.add().basic
    expression.logical_operator = LogicalOperator.AND

    expression.comparisons.append(ComparisonOperator.GREATER_EQUAL)
    expression.operands.add().Pack(Test(attribute1=1, attribute2=1, attribute3=True))

    expression.comparisons.append(ComparisonOperator.LESSER_EQUAL)
    expression.operands.add().Pack(Test(attribute2=10, attribute2=10, attribute3=True))

    def yield_search():
        yield search

    for response in roguedb.search(
        yield_search(),
        metadata=[("authorization", f"Bearer {encoded_jwt}")]):
        results = []

        # Queries are zero-indexed. Partial results get sent
        # and mapped to that index.
        for result in response.results[0].messages:
            test = Test()
            response.results[0].Unpack(test)
        
        # Each response sends a list of the query ids finished
        # with processing.
        if 0 in response.finished:
            pass # Finished processing.

    # Example of a basic non-indexed query. 
    # Search Query: Test.attribute1 < 1 AND Test.attribute2 != 10
    search = Search(api_key=api_key)
    expression = search.queries.add().basic
    expression.logical_operator = LogicalOperator.AND

    expression.comparisons.append(ComparisonOperator.LESSER)
    expression.operands.add().Pack(Test(attribute1=1))
    expression.fields.append(1) # Corresponds to field id in test.proto

    expression.comparisons.append(ComparisonOperator.NOT_EQUAL)
    expression.operands.add().Pack(Test(attribute2=10))
    expression.fields.append(2) # Corresponds to field id in test.proto

    #####################################
    ##### Schema Change API Example #####
    #####################################

    subscribe = Subscribe(api_key=api_key)

    # All proto files should be sent in a list of
    # their contents. No modifications required.
    for file in detect_files(directories=[
        "absolute/path/to/protos/directory1",
        "absolute/path/to/protos/directory2"]):
        with open(file) as input_file:
            subscribe.schemas.append(input_file.read())

    # Any schemas excluded will have associated data deleted.
    # Schema change failure results in no changes applied.
    response = roguedb.subscribe(subscribe)
